src/stateStreamRegistry.ts

```diff
@@ -0,0 +1,87 @@
+import { PersistentReducedStateStream } from './PersistentReducedStateStream';
+import { ActionStream } from './types/helpers';
+
+export class StateStreamRegistry {
+  streams: Map<string, PersistentReducedStateStream<any>> = new Map();
+  /**
+   * The action stream and states the registry has been started with, if started.
+   * Should be `null` before `startReducing` and after `stopReducing` has been
+   * called.
+   */
+  started: { action$: ActionStream; states: Map<string, any> } | null = null;
+
+  /**
+   * Register a stream to this stream registry
+   *
+   * If a stream with the same name has already been registered, an error is
+   * thrown.
+   *
+   * If the registry has been started, the registered stream will also be
+   * started.
+   *
+   * @param stream The stream to register
+   */
+  register(stream: PersistentReducedStateStream<any>) {
+    if (this.streams.has(stream.name)) {
+      throw new Error(`Duplicate stream name: ${stream.name}`);
```

Comment:
From testing a bit locally, this is actually a bit tricky to achieve 🤔 

When `branchId$` is hot-reloaded, the only way to "dispose of it" is from within `branchId$` so it means that every state stream file would need unregistering logic (not ideal).

One alternative is to handle `re-registering` of a stream as hot-reloading, but this can lead to problems if you inadvertently register two streams with the same name.

I.e.:

```typescript
  register(stream: PersistentReducedStateStream<any>) {
    if (this.started !== null) {
      const duplicateStream = this.streams.get(stream.name);
      if (duplicateStream) {
        duplicateStream.stopReducing();
      }
      stream.startReducing(
        this.started.action$,
        this.started.states.get(stream.name)
      );
    }
    this.streams.set(stream.name, stream);
  }
```

^ this doesn't feel right, and it will cause problems if you have duplicate stream names, but I think it solves the hot-reloading issue :P

Reply:
After some thinking, an alternative solution for the duplicate problem could be to only throw an error if the registry is not started yet. 

Then it becomes:
```typescript
  register(stream: PersistentReducedStateStream<any>) {
    const previouslyRegisteredStream = this.streams.get(stream.name);
    if (this.started !== null) {
      let state = this.started.states.get(stream.name);
      // Since we've already started the stream registry
      // assume that we are "re-registering" the stream due to hot-reloading
      if (previouslyRegisteredStream) {
        state = previouslyRegisteredStream.state;
        previouslyRegisteredStream.stopReducing();
      }
      stream.startReducing(this.started.action$, state);
    } else if (previouslyRegisteredStream) {
       throw new Error(`Duplicate stream registered before starting the stream registry. name: ${stream.name}`);
    }
    this.streams.set(stream.name, stream);
  }
```

This seems to solve the hot-reloading problem and the duplicate stream problem if my assumptions are correct :)

Reply:
Thanks for looking into this. I didn't have the chance to test hot reloading before publishing :/

I like the idea of allowing re-registration only after the registry has started, but I also don't quite like the state dependent behavior.
Another approach could be to handle this differently depending on dev/prod environment. So in prod we would always throw an error, while in dev we would always re-register, perhaps with a `console.warn`.
`tsdx` (which we are not, yet, using) has proper support for [separate dev and prod bundles](https://github.com/jaredpalmer/tsdx#development-only-expressions--treeshaking), so I think this might be a viable approach.